### DISCUSSION TITLE
Plans for creating a Range class/extension type comparable to Rust's Range structs?

### DISCUSSION THREAD

**Discussion by cds-reis** [mcmah309/rust#5] [Ideas]
With the new addition to the range(int start, int end) function, I was wandering if there are any plans to replicate Rust's Range structs (https://doc.rust-lang.org/std/ops/index.html#structs) to the project, with the options to create different types of Ranges (differentiating between inclusive or exclusive, and bounded or unbounded), and using it on Iterables to create Slices (I know Slice already has a way to do it, but AFAIK only with indexes).

If there isn't any plan, I wouldn't mind doing it myself btw.

**Comment by mcmah309** [comment]
I considered it, but I couldn't define an api that I considered idiomatic for Dart.
Something like
```dart
for(final x in RangeInclusive(0,10)) {

}
for(final x in arr(RangeInclusive(0,10))) {

}
```
Didn't sit right with me. While
```rust
for x in 0..=10 {

}
for x in &arr[0..=10] {

}
```
Is more natural for Rust. 
This felt more natural for Dart
```dart
for(final x in range(0, 10 + 1)) { // just an example, 11 is obviously preferred

}
for(final x in arr.slice(0, 10 + 1)) {

}
```
I also explored adding inclusive and exclusive to `range` e.g.
```dart
Iterable<int> range(int start, int end, {bool sInc = true, bool eInc = false}) {
  if (!sInc) start += 1;
  if (eInc) end += 1;
...
}
```
But this doesn't really add anything IMO because setting `eInc = true` is the same adding `1` to `end` as previously shown.

Open to differing opinions though, and If you have an idea for an api and it flows well with the language, definitely interested in hearing as well!

**Comment by mcmah309** [comment]
After sitting on it for a bit longer, I don't see a reason not to add it. User choice. If you want to work on something I'm for it!

**Comment by mcmah309** [comment]
I had some time today so I ended up implementing it! https://mcmah309.github.io/rust_core/libs/ops/ops.html#rangebounds

**Comment by cds-reis** [comment]
Yeah I like this implementation. I also wish there was some extensions on int to make creating ranges easier, like:
```dart
import 'range.dart' as r;

extension RangeExtensions on int {
    Iterable<int> range(int end) => r.range(this, end);
}

void main() {
    for (int index in 0.range(10) {
        // ...
    }
}
```
I would also be weary of the current implementation of the `range` function, since it gives a bounded range, even if sometimes you would like for it to be unbounded. I would advise to let the default range be a strict `range(int start, int end)`, and then create variations from it, to avoid creating some confusion.

**Comment by mcmah309** [comment]
As a matter of opinion, since `0.range(10)` has the exact same number of characters of `range(0,10)` (could also be `range(10)`), I don't find adding another variation compelling. But certainly if you like it, go for it!

More often than not, we reach for bounded ranges than unbounded ones, thus that became the default. Which is the same behavior in python. If an unbounded range is desired, one could
```dart
for(int x in RangeFrom(0)){} // 0, 1, 2, ...
for(int x in RangeFrom(0).iter().stepBy(10).map((e) => -1 * e)){} // 0, -10, -20, ...
```
Maybe something can simplify the second case though, but all the `RangeBounds` classes have the same behavior as the Rust ones. The `Range` class fulfills `range(int start, int end)`.
